Erase regions when confirming transmutability candidate

compiler-errors opened the pull request:
Fixes an ICE where we call `layout_of` on a type with infer regions.

compiler-errors committed:
Erase regions when confirming transmutability candidate

petrochenkov commented:
r? @scottmcm to balance https://github.com/rust-lang/rust/pull/110021#issuecomment-1500187366

scottmcm commented:
This looks plausible to me -- not having the `Binder` in the `is_transmutable` is a nice change -- but I don't think I know enough about what's going on here to be confident in signing-off.

r? types

spastorino commented:
r? types

lcnr reviewed (commented):
nit, then r=me

compiler-errors commented:
Addressed the comment nit and rebased past a PR that changed the transmutability API slightly (no more binder for an argument).

We lose a `'static` in an error message, but that's not particularly important IMO, since again, lifetimes don't affect transmutability.

@bors r=lcnr rollup

bors commented:
:pushpin: Commit f6bfb4bf8e62e5f924790e62b4e8321b7658da2b has been approved by `lcnr`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=lcnr f6bfb4bf8e62e5f924790e62b4e8321b7658da2b -->
<!-- homu: {"type":"Approved","sha":"f6bfb4bf8e62e5f924790e62b4e8321b7658da2b","approver":"lcnr","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
